content/advent-2019/pkgapi.md

```diff
@@ -0,0 +1,1922 @@
+
+The use of the terms "static" and "dynamic" is not good, but are in standard use. To make it clearer, allow me to further explain:
+
+A resource is static if its state is known at compile time. A resource is dynamic if its state unknown at compile time.
+
+Thus the Shakespeare resource is static because the entire corpus is known and available at compile time. The state of a graphics card availability may change and needs to be determined at runtime, therefore it is a dynamic resource.
+
+At the time of writing, there is [a proposal to allow static resources to be embedded in the final binary of a Go program](https://github.com/golang/go/issues/35950), so the story is still to be told on the Go end.
+
+# The Types of Libraries #
+
+Barring discussions on the concrete details of libraries and what form they take (packages, drivers, modules, etc), let's consider the various types of libraries we have discussed so far.
+
+Broadly speaking there are two general classes of libraries:
+
+* Libraries where source code is the primary resource that is being shared.
+* Libraries where some resource other than source code is being shared.
+
+The latter of these can be split into two:
+
+* Driver libraries
+* Resource libraries
+
+A driver library is typically a package that wrap access to a driver. For example, the aforementioned  [`cu`](https://gorgonia.org/cu) package wraps the CUDA drivers to enable CUDA programming in Go. Similarly the [go-gl](https://github.com/go-gl) package which mediates access to OpenGL). Both packages expose extra helper functions to make the transition between worlds easier, but are fundamentally drvier packages.
+
+The `willshakes` library is an example of a resource library. For a real life use case, I offer the [mnist](gorgonia.org/gorgonia/examples/mnist) library for consideration. Due to the nature of the source data, as well as the inability for Go to handle static resources, the design of the package is limited to source code that loads data from an external file into a data structure. In the Python world, this is a different case as one may use [`keras.datasets.mnist`](https://keras.io/datasets/) immediately as a resource.
+
+
+
+
+
+# What Makes A Good Library? #
+
+Now that I have defined the things that can go into a library, let's take a look at what makes a good library. We will start with the simpler, more obvious statements, before moving on to more nuanced considerations. Despite this, the simpler statements often come with caveats, which will be briefly explored.
+
+I have a few principles that form the basic principles of what makes a good library:
+
+1. Reliable
+2. Easy to use/build
+3. Generic
+
+## Reliable ##
+
+First and foremost, a library must be reliable. What is the point of using a library if it cannot reliably do what it claims to do? There are many features of a  reliable library, enumerated below.
+
+### Does One Thing ###
```

Comment:
I usually prefer wording "has a clear value proposal", rather than does one thing. Packages such as `http`, clearly do not do one thing, however it does have a clear purpose.